### ISSUE TITLE
Compatibility with new Firefox Sync

### ISSUE THREAD

**Issue by ogasser** [owncloud/mozilla_sync#33]
The [Sync Storage API v2.0](https://docs.services.mozilla.com/storage/apis-2.0.html) is in **final draft** status. According to the [roadmap](https://wiki.mozilla.org/Identity/Roadmap) we could see them incorporated into the Mozilla clients in early to mid 2014. To prepare for this change we need to start implementing the new API before it is actually deployed to all Firefox clients.

Also, Firefox needs to allow defining a custom sync server. We opened this issue for that in the Mozilla bugtracker: https://bugzilla.mozilla.org/show_bug.cgi?id=989756 – please follow that.

**Comment by godfuture** [comment]
Any updates on progress?

**Comment by nickel-chrome** [comment]
I've updated [weaveclient-python](https://github.com/nickel-chrome/weaveclient-python) to support the v1.5 protocol, which should hopefully help with development of the server. But not sure where server itself is up to.

**Comment by posativ** [comment]
@nickel-chrome how much code from your client can be shared with e.g. https://github.com/posativ/weave-minimal?

**Comment by nickel-chrome** [comment]
Not much I'm afraid.

On the other hand I don't think there is actually that much development to do. The hard stuff is largely handled by [PyBrowserID](https://github.com/mozilla/PyBrowserID) and [PyFxA](https://github.com/mozilla/PyFxA). And having a working client *should* make it much easier to develop a server.

A starting point for BrowserID authentication on the token server could be this function from [weaveaccount.py](https://github.com/nickel-chrome/weaveclient-python/blob/master/weaveaccount.py):

```python
def verify_assertion(audience, assertion, local=True):

	if local:

		v = LocalVerifier(["*"], warning=False)
		result = v.verify(assertion)

	else:
		url		= "https://verifier.accounts.firefox.com/v2"
		payload = json.dumps({"audience": audience, "assertion": assertion})

		req = urllib2.Request(url, data=payload)
		req.add_header("Content-Type", "application/json")
		req.get_method = lambda: 'POST'

		try:
			f = opener.open(req)
			result = json.loads(f.read())
			f.close()

		except urllib2.URLError, e:
			raise WeaveException("FxA sync auth token request failed: " + str(e) + " " + e.read())

	if (result['status'] == 'okay'):
		return True
	else:
		logging.debug(pprint.pformat(result))
		return False
```

Similarly a starting point for Hawk authentication on the 1.5 storage server could be this function from [weaveclient.py](https://github.com/nickel-chrome/weaveclient-python/blob/master/weaveclient.py):
```python
def verify_auth_header(self, method, url, data, header):
    logging.debug("verify_auth_header()")

    parsed_url = parse_normalized_url(url)

    req = {
        'method': method,
        'url': url,
        'host': parsed_url['hostname'],
        'port': parsed_url['port'],
        'headers': {
            'authorization': header
        }
    }

    logging.debug("hawk auth request:\n" + pprint.pformat(req))

    credentials = {
        self.id.replace('=', ''): {
            'id': self.id,
            'key': self.key,
            'algorithm': 'sha256',
        }
    }

    logging.debug("hawk auth credentials:\n" + pprint.pformat(credentials))

    server = hawk.Server(req, lambda cid: credentials[cid])

    # This will raise a hawk.util.HawkException if it fails
    artifacts = server.authenticate({})

    #logging.debug("hawk auth artifacts:\n" + pprint.pformat(artifacts))

    return True
```

**Comment by nickel-chrome** [comment]
@posativ also as @ockham has [noted](https://github.com/owncloud/mozilla_sync/issues/33#issuecomment-58814511) there are reference implementations of both the token server and the v1.5 storage server, from which hopefully the core authentication code can be better understood and possibly reused.

**Comment by kodejak** [comment]
The Mozilla SVN is public since a bug in 2013. I hope this helps.
https://viewvc.svn.mozilla.org/vc/

**Comment by allo-** [comment]
Does anybody know, how much/how complicated code weave 1.1 is?
Maybe it would be an easy and clean solution to create a weave-addon implementing the 1.1 protocol for current firefox versions.

The whole 1.5 does very similiar things on the storage backend anyway and just integrates better with the firefox accounts stack, simplifies setup (at the cost of some security) and has some minor improvements / new techniques (like hawk authentication) added.
So sticking with a sync 1.1 addon may be a good choice for many users, if it's possible to maintain one.

**Comment by posativ** [comment]
The server-side code is much simpler than the client side code. The client side handles the crypto, retrieves different buckets etc. Also, a Firefox addon must be actively maintained because Firefox itself has a quite fast release cycle.

**Comment by yonas** [comment]
For some reason I got the impression that mozilla_sync would not work with the latest version of Firefox. It seems to work well. Tested with Firefox 37 on Ubuntu 15.04 and Firefox 37 on Windows 7.

Wasn't able to get it to work with Firefox on Android (4.4.2) though.

**Comment by meeotch** [comment]
Argh.  Just went through the whole process of installing, setting up, and syncing before discovering this issue.  Between the petering out of this thread, and the one at bugzilla (https://bugzilla.mozilla.org/show_bug.cgi?id=989756), sounds like the end of the line for this app?

**Comment by yonas** [comment]
@meeotch Like I said, I didn't find any issue here. You might want recheck your configuration steps.

**Comment by allo-** [comment]
With recent developments, Full-Stack deployment will become less usable:
Firefoxaccounts will be used for sync, Firefox Apps, Pocket (newest integration) and more Mozilla services, where the Accounttoken will only be accepted, if it's from the Mozillaserver. 

Most probably the whole Firefoxaccounts are thought as central account for purchasing webapps and apps for firefox os devices.

The other way round, deploying only the syncserver-Part has the problem, that your syncserver has to accept tokens from your mozilla login and the own syncserver adds little security.
The problem there is, that the login happens on a mozilla site (iirc clientside per javascript, but JS delivered from the mozilla server). This means the mozilla server may log your password on login and then access the syncdata (if it happens to know your server url, which probably is not transfered to mozilla).

This looks like it's either only worth to implement the sync part (not the account part) or try to backport old sync to new firefoxes (i.e. palemoon seems to maintain sync 1.1 in its firefox fork).

**Comment by meeotch** [comment]
I should clarify:  I didn't mean that it doesn't work with the current version.  I meant that it sounds like there's not enough support on the mozilla side or on this side to make it work with future versions.  From the bugzilla thread:  "The upshot is that Firefox engineering will not support Sync 1.1 clients past about Firefox 39; and that Firefox Cloud Services will not be running Sync 1.1 infrastructure past about August."

So - end of the line, sooner than later.  (For me, sooner - since Android Firefox already can't sync with owncloud, and I have more Android devices than desktop installations.)

**Comment by Armada651** [comment]
> This means the mozilla server may log your password on login and then access the syncdata (if it happens to know your server url, which probably is not transfered to mozilla).

Such an access to sync data will also be logged on the sync server, so it's not a very good way to do surveillance without anyone being able to find out.

**Comment by allo-** [comment]
Every login from another person may be logged. This does not change the problem, when somebody logs my accountpassword during login and fetches my saved passwords from my own syncserver.
The whole reason for full-stack is avoiding others to be able to get your login credentials.

**Comment by Armada651** [comment]
While I am a big proponent of a full stack solution, you shouldn't save your server password in Firefox.

**Comment by allo-** [comment]
i speak about the whole password safe. You might argue password safes are no good idea, but weak passwords are even less. But this is not even the topic here, i might find my bookmarks, history, ... private as well.

**Comment by Armada651** [comment]
I wasn't arguing that, but if you assume someone is able to access your password safe by compromising your sync account you will have to review the security of your password safe.

Sure, I also consider those highly private, but like I said using your credentials to log in to your sync server is not a very good way to invade your privacy without it being logged.

**Comment by allo-** [comment]
I care about being hacked, disregarding if the hacker gets logged.

Now: I have my sync 1.1 server, firefox uses it, nobody knows where it is, it uses https, the key-value pairs are encrypted with a key stored locally on each pc i am using.

Sync 1.5:
Login on Firefox-Server, which encrypts the password (possibly grabbing it there),
then uses the encrypted password to login to the account to get a token.
Then the token authenticates me to my own sync 1.5. server, where i get the key-value pairs encrypted with the password (not the encrypted password, which was sent to the mozilla server).
Security issues: The login happens on a website, not in the firefox chrome, which means there may happen stuff like JS injection, which is not controlled by me. That grabbing the password is enough to get anything can be seend without reading source or long documentation by coupling a new device to sync without knowing more than the email address and the password.

Full-Stack:
I use my own accounts server and secure it myself, that nobody other than me can grab the password on the login page. Place of loginserver and syncserver are unknown to most people (except somebody reading your traffic).

Still less secure than 1.1, where the sync-key never leaves the browser chrome. I would prefer the whole login to happen inside a firefox dialog instead of on a login-webpage, which is just integrated into the gui via a frame.

So i wait for some good full-stack solution. Or somebody putting sync 1.1 into a firefox extension. the full stack with mozilla tools is a real big stack, so a owncloud plugin, which manages to bring it down to the essential stuff would be good.

**Comment by allo-** [comment]
Btw: I seems, that you cannot couple new devices to sync 1.1 with firefox 38. You can still get the dialog, if you create a key ``services.sync.username`` in about:config with a nonempty string, but when clicking okay i cannot see a request in my server log.

**Comment by Armada651** [comment]
I also find it laughable that a company who triumphs is securing our privacy would make it harder and harder to keep control of our data.

**Comment by prometheanfire** [comment]
seems like the api spec is fairly doc'd out at least

**Comment by jknockaert** [comment]
At one time Mozilla planned to draw on the owncloud experience with Sync 1.1 to get an easy self-hosting package for 1.5: https://blog.mozilla.org/services/2014/05/08/firefox-accounts-sync-1-5-and-self-hosting/
Not sure what happened after that.

**Comment by allo-** [comment]
you can see my comment there, its a PITA.

Here the best approach for their software, even the author of the post edited, that the howto is now broken after mozilla changed stuff again:
https://thomas-leister.de/internet/ubuntu-firefox-sync-1-5-server-installieren/

I do not know if it would work with current firefox and older mozilla-serverstuff, but i had this for a short time running and then it broke without any obvious reason.

I already had a look on the API, it looks not too complicated, despite of using HAWK, which i never used before.

**Comment by allo-** [comment]
The thomas leister howto is now offline. I guess he gave up to update his howto.
Here is a new url (untested): https://www.pixelcontainer.net/firefox-syncserver-1-5-installieren/

**Comment by coder9** [comment]
There's a copy on Archive.org from March last year...

https://web.archive.org/web/20150323050141/https://thomas-leister.de/internet/ubuntu-firefox-sync-1-5-server-installieren/

Maybe if someone could translate it into English that would help other contributors?

**Comment by nickel-chrome** [comment]
Unfortunately a fair bit of work is required to make mozilla_sync compatible with the new sync v1.5 protocol. In short mozilla_sync will need implement the Token Server API v1.0 in addition to the SyncStorage API v1.5.

The key change is that authorization of requests has changed from Basic Auth to Hawk Auth, this is to ensure that the password never leaves the client device, which is a necessary security measure as with sync v1.5 the sync key can effectively be derived from the password (its a fairly involved process to ensure that neither Mozilla nor the sync server host see the password).

I would really love mozilla_sync to be compatible with Sync v1.5, and at some stage I may well find the time to do this myself, but this could be some time off :( On the other hand if anyone wants to have a go, I now understand the new Token Server API and Hawk Auth quite well and would be more than willing to give a hand.

Disclosure: I've setup a sync server at cucumbersync.com to support and end-to-end encrypted contacts and calendars app for Android. If this is of interest the latest beta release of Cucumber Sync can be found below. All going well it will be released on Google Play and as a browser extension 'real soon now'.

https://github.com/eXfio/CucumberSync/releases

Cucumber Sync is built on Firefox Sync and as such I've got Mozilla's standalone 1.5 sync server running a token server endpoint at https://cucumbersync.com/syncserver/token/1.0/sync/1.5, which you are welcome to try out. I am also happy to help others set up their own. Its not too hard, but your web server does need to support wsgi and for Apache I found I needed to edit the main Apache config file, the config is not supported in .htaccess config.

**Comment by allo-** [comment]
Untested, but promising: https://github.com/mozilla/fxa-local-dev
Seems they have there a meta-project, which downloads and runs all needed projects for a development server (self hosted server).

**Comment by josh4trunks** [comment]
anyone want to setup a bounty source for this? might be a good way to add incentive to get this worked on. I bet we could get this job to be worth it for someone.

**Comment by Armada651** [comment]
Honestly I think the reason no developer wants to work on this is because the future of self-hosted FF sync is very unclear. Firefox doesn't even provide an UI to specify a self-hosted server anymore. Any developer looking at this will conclude that Mozilla is dropping support for self-hosted servers. Which makes it highly likely that all hours invested in this will be wasted when Mozilla finally drops the hammer on this feature.

**Comment by allo-** [comment]
They will not drop the feature. Firefox is opensource and the developers need the option to use a testserver.
But the feature will stay unaccessible and the server is a pain to set up. So self-hosting is de-facto dead.
Someone should make sync 1.1 into an addon, as there are several server implementations for it. Or start a totally new sync addon. Like foxmarks was, before they dropped self-hosting support as well.

**Comment by Armada651** [comment]
Having an add-on that provides an easy-to-use UI for the self-hosting features would go a long way to incentivize contributors to work on features such as this. Ofcourse even better would be to contribute a patch to Firefox to provide an official UI for it again, however unlikely that may be.

**Comment by jonas2515** [comment]
As far as I know it also is really difficult to set a custom sync server on android, because there's no about:config value for it (I don't know if it changed, though). Actually firefox has the same support for custom sync servers like chromium at the moment...

**Comment by coder9** [comment]
The problem with using sync 1.1 or building a sync 1.1 add-on is the 1.1 protocol jumbled the bookmarks up if you added Android clients.

Perhaps Vivaldi might be more open to allowing a custom sync server. They are working on their sync functionality now so it might not be too late for them to support custom servers properly.

**Comment by Armada651** [comment]
Isn't it also possible to build a sync 1.5 add-on that provides an UI for it? If the self-hosting features are easy-to-use there will be much more incentive to update the owncloud plugin.

**Comment by jonas2515** [comment]
@coder9 Vivaldi is closed source, I don't think they'll publish their sync server code.

**Comment by beyerservice** [comment]
@jonas2515 there is an addon for android: https://addons.mozilla.org/de/android/addon/fxa-custom-server-addon/
But I am no developer and I cant to this for Firefow for Desktops...sad :-(

**Comment by Larx** [comment]
Is it possible that finally with Firefox 42.0 old-style sync is gone :( ?

**Comment by franzhartwig** [comment]
Running Firefox 42.0 on Ubuntu and Windows 7, syncing with Owncloud 8.1.3 and Mozilla_Sync 1.4. It works.

**Comment by Larx** [comment]
Yes, false alarm. I played around with password and bookmark sync alternatives, and somehow some KeePass-plugin did not uninstall cleanly and caused the issue. Sorry for the noise.

**Comment by mlan** [comment]
I have the same setup as franzhartwig and sync is working (both with win7 and android). I have not seen any reports that mozilla_sync 1.4 is working on owncloud 8.2.1. I'm waiting to upgrade my server since I do not want to risk loosing mozilla_sync.

Have anybody have mozilla_sync 1.4 is working on owncloud 8.2.1?

**Comment by Ruud68** [comment]
works for me: Ubuntu, firefox 42, ownCloud 8.2.1 (stable),  Mozilla Sync 1.4

**Comment by mlan** [comment]
@Ruud68, Great news! :) Many thanks!

**Comment by Larx** [comment]
Addon is still working flawlessly for me on current OC, too. As the old-style sync is scheduled to get removed from the Firefox code base soon, this won't help too long, I'm afraid.

**Comment by mach-o** [comment]
I wasn't able to get Firefox 43 to work, so downgraded to 42. If you had better luck, please let me know.

**Comment by allo-** [comment]
FF43 works with sync 1.1 (and django syncserver) for me.

**Comment by JB1985** [comment]
Master mozilla_sync with Firefox 43 and owncloud 8.2.1 not work!

**Comment by 1of16** [comment]
it is still working with FF43.x and owncloud 8.2.x

**Comment by RealRancor** [comment]
Same here, also still works with FF 43.x and 8.2.x

However it really makes no sense to post in here if it works or not as the old app is a dead end.

**Comment by 1of16** [comment]
mozilla announced, that the support for the "old sync" will be discontinued with FF 30....and its still working.
so I guess this app is no dead end and there is _no_ alternative atm (not even with the glorious chrome).

**Comment by allo-** [comment]
It's not that dead, as palemoon sticks to old sync.

And there are big changes in firefox ahead*, which might lead to even more forks.

* Removal of XUL/XPCOM and a lot of extensions, which will not work without.

**Comment by RealRancor** [comment]
Palemon is not a current FF 43. ;-)

**Comment by rnewman** [comment]
We plan to remove the legacy Sync account type from Android devices in Firefox 46: https://bugzilla.mozilla.org/show_bug.cgi?id=1220906. I expect desktop will be in roughly the same timeframe: the Mozilla Sync 1.1 servers were decommissioned a while ago.

**Comment by Ruud68** [comment]
Hi, I personally think this is a loss. In the current times of the 'almighty' google pushing their chrome browser it is very important to team up. Owncloud is really getting lift-off with their fabulous solution so being able to use their solution will give a competitive edge and will be beneficial for both owncloud and firefox.
Just my thoughts :)

**Comment by 1of16** [comment]
as far as I see it there is atm _no_ alternative to the owncloud-sync.
It would be nice to know in which desktop-versions (normal and esr) it will be removed and which "alternative" firefox browsers (waterfox etc). maybe still able to work with the owncloud-sync.

nevertheless: it is a bad move to remove the support and don't provide an alternative.

**Comment by Feandil** [comment]
> as far as I see it there is atm no alternative to the owncloud-sync.

There is the 1.5 sync server published by Mozilla: https://docs.services.mozilla.com/howtos/run-sync-1.5.html (that need https://docs.services.mozilla.com/howtos/run-fxa.html if you want to be completely separated from their servers), right?

**Comment by allo-** [comment]
I tried to install it. The sync server is pretty easy, the accounts server is very very complicated. Once it run, then it broke and after that i never got it working again. The documentation is sparse, the stuff seems bloated, the first "node install" pulls like hundreds of dependencies and the configurations are a bit confusing, including stuff like authentication tokens for the official server in the source code and own certificate store you need to patch to get cacert certificates accepted, which are installed in the linux system.

Nothing as easy as the owncloud or django-weave servers. I really wish somebody uses the mozilla sync-client code to make a sync 1.1 addon. But as mozilla will cease to support powerful addons and migrate to webaddon-api similiar to chrome, it's not likely such an addon will be possible in the future.
I read a bit in the current restartless-addon docs for looking into an own simple bookmark/history sync addon with easy serverside, but it seems to lack a history-api.

**Comment by rnewman** [comment]
Correct — if you want to be *entirely* self-hosted, you'll need to host FxA, too, and those are the two docs.

Hosting your own FxA has some tradeoffs, not least of which is confusion about which account you're using as you interact with the browser and web content: for example, I think Bugzilla is going to start supporting FxA as a login method, and I doubt it'll trust a self-hosted FxA, even if that's what you're using with your own Sync server.

> nevertheless: it is a bad move to remove the support and don't provide an alternative.

We shipped FxA Sync to release almost two years ago. We've been in the process of gradually EOLing Sync 1.1 ever since. Documentation on how to self-host FxA Sync has been around for at least a year, which is a pretty long overlap for people to get their shit together.

If there's not enough of an ownCloud developer community to keep Firefox Sync support current, given a two-year window, then I'd think very hard about whether I would use ownCloud for Sync hosting.

(Of all the things to use ownCloud for, Sync is the least important, IMO. Client-side-encrypted storage hosted by Mozilla is very much at the trustworthy end of the spectrum, and I'm very happy letting Mozilla's operations team handle durability and availability for me.)

It does not make sense for the Mozilla community to continue to support Sync 1.1 just because ownCloud hasn't maintained protocol compatibility. That's a bad tradeoff.

**Comment by rnewman** [comment]
> If there's not enough of an ownCloud developer community to keep Firefox Sync support current, given a two-year window, then I'd think very hard about whether I would use ownCloud for Sync hosting.

I should clarify: that's not dissing ownCloud developers; simply recognizing that in a world with limited time to do work, picking what to spend that limited time on is important.

**Comment by allo-** [comment]
Nope, FxA was up to now never easy to setup. I really tried it, it seems to be written exactly to run for mozilla.
Here are some old docs, i think the author deleted the post, because he stopped trying himself and the people in the comments were coming back with issues all the time.

https://web.archive.org/web/20150323050141/https://thomas-leister.de/internet/ubuntu-firefox-sync-1-5-server-installieren/

I think FxA is a bad move, as it goes from decentral concepts like Weave, OpenID or what Personas wanted to become (and never was) to a very centralized concept. And it seems, that Mozilla wants to use FxA for some webapp-Store maybe even for firefoxos, if something of it remains (is it now dead or is not?). And this will not accept own accountservers (which would not be a problem with a protocol like openid), neither will bugzilla, i guess.

Another problem is, that the needed settings are really hidden and unclear. And last but not least it seems that the whole login-process is not that secure as claimed. I did not really trace it in the source, but the login page is loaded from the server hosting FxA and not a chrome-page, so if it's highjacked any "we hash before" javascript can be replaced with a "we send your password home" javascript. And you can login everywhere just with e-mail and password, which means who's phishing e-mail and password will be able to login.
I guess that's not too harmful, as you only need to login once per device, though.

Maybe the local-dev project i posted before would be useful for getting the server side running. But this is of course out of scope for owncloud, which needs to implement it by itself.

**Comment by jknockaert** [comment]
@rnewman The originial plan for Sync 1.5 was:
>Lastly, we need to ensure that the work we do to facilitate self-hosted Sync is the right work. Namely, how do we go from “technically possible” to “reasonably easy” for advanced users? To get this right, we’ll need to draw on the experience of projects like ownCloud, which is at the forefront of making Sync 1.1 easy to self-host, as well as users who have self-hosted in the past or who are interested in self-hosting now.
(from https://blog.mozilla.org/services/2014/05/08/firefox-accounts-sync-1-5-and-self-hosting/)
It seems a reasonable expectation of Sync 1.1 users to stick to the ownCloud platform. And there must be more than a couple who've chosen this solution: even today the Mozilla Sync 1.4 app is still in the top 10 of downloads in the ownCloud app store. So if Mozilla doesn't care now to leave ownCloud hosted Sync 1.1 users behind, it's more like Mozilla changed their mind but don't blame these users.

**Comment by rnewman** [comment]
> The originial plan for Sync 1.5 was…

No, the original plan for Sync 1.5 was to ship a Firefox Account-authenticated Sync implementation on two platforms as quickly as possible, with a staged transition period for users, a longer "you should jump" period for self-hosters, and a ramp for self-host documentation from "none, sorry" to "good enough".

Various people have/had various responsibilities in that; obviously a blog post about self-hosting will come at that from the self-hosting perspective.

> It seems a reasonable expectation of Sync 1.1 users to stick to the ownCloud platform.

I think you might be misunderstanding me, and misunderstanding whatever doc you quoted that from.

We'd be very happy for ownCloud to continue to support Firefox Sync. It just can't keep doing it with Sync 1.1, because Sync 1.1 in Firefox (a) isn't free to maintain, and (b) isn't well tested, and thus will bitrot. Sync 1.1 is dead, and its body needs to be buried. You can totally keep running ownCloud's Sync app as it currently stands… but later this year none of your clients will sync to it.

That quote is saying that any work done on self-hosting for Sync 1.5 should push towards making things easier. It doesn't imply maintaining legacy protocol support, and it doesn't imply taking responsibility for ownCloud.

**Comment by allo-** [comment]
Ever since FF 29(?) i was always testing with a copy of my profile, if sync still works. This is important for us and sync 1.5 is neither well supported for self hosting on the serverside, nor we can see, that its supported on the client side, as there is ABSOLUTELY NO gui to expose it. Sync 1.5 seems not too bad, but Firefox should REALLY decouple it from FxA.

At least i now know, i can upgrade without doubt up to FF45.

**Comment by craigarno** [comment]
Okay, my 2 cents... or 0.02 euro...

I would like to see a Firefox feature to disable "upgrades" past FF45.  I do not want to keep following Mozilla upgrades until I can find a reasonable replacement for Sync 1.1/OwnCloud Mozilla_Sync 1.4.  Currently Mozilla's Sync 1.5 doesn't look like a "reasonable" replacement to me for all the reasons mentioned earlier; FXA too complicated to implement, Mozilla self hosting solution too complicated to install due to a large number of dependencies, inability to keep the Mozilla self-hosting solution running, and I'll add my own point which is I need to add (and support) even more things to my web server to support Sync 1.5 using something other than PHP/Ajax, which is what OwnCloud supports.

Sync 1.1 (Weave) continues to work well for me, without problems on about 7 computers I currently use.  My solution will be to stop allowing Firefox upgrades (which seem almost mandatory) at the point where Sync 1.1 client support is removed.

Other than this, yes, I understand the effort involved in supporting sync 1.1 (weave) code.  What I feel is being ignored is the amount of user effort required to follow Mozilla's FxA/Sync 1.5 agenda, regardless of allowing 1-2 years for "other developers" to "get with the program" of "Mozilla's agenda".

What you're hearing, and trying to ignore as "inconvenient" is users don't want a change which precludes the use of a working and trusted Sync 1.1 solution.

Believe it or not, I have a lot of other hardware and software development activities designing medical equipment which saves peoples lives.  I can't currently afford to stop that to develop support for Sync 1.5, which last time I looked, looked like development would consume a lot of time I can't provide right now.  So my only solution is to reject Mozilla Firefox upgrades.

**Comment by beat** [comment]
My 2 cents: Bottom line is that if Firefox wants to seriously continue to promote Freedom for your users' data, it should make self-hosting sync implementations much easier (e.g. support alternative simpler synchronization server protocols and cooperate with e.g. OwnCloud, e.g. contributing a compatible Sync plugin).

Owncloud users represent maybe not a big market-share compared to "normal" Firefox users. But they are all advanced evangelists of Freedom of user data, and that should count tenfold for Mozilla. Don't let OwnCloud users and other free platforms fall... Proposal: Support 1.1 until a simpler 1.5 version protocol and OwnCloud plugin is done/contributed.

**Comment by mach-o** [comment]
Craig: Fair enough, but keep in mind that if your goals are trust and security, then it’s also risky to cling to software that receives no security updates, even after vulnerabilities and exploits are discovered.

**Comment by HLFH** [comment]
FYI, there is a great alternative to Firefox Sync 1.5 server: [wallabag](https://github.com/wallabag/wallabag)

With wallabag, you bookmark here not only the URL but also the content of the page. You could [create an account on Framabag](https://framabag.org/) to test the v1.9.1 if you don't want to install it on your own server.

 Chrome and Firefox extensions, and also Android, iOS, Firefox OS, Windows Phone apps [are available](https://www.wallabag.org/pages/download-wallabag.html).

There is even also a [French wikipedia page](https://fr.wikipedia.org/wiki/Wallabag) related to this open source project. 

Maybe you should switch from Firefox Sync 1.1 or Sync 1.5 to Wallabag. That what I've done and I'm really happy with it. I'm using v2 alpha 1 and it works like a charm.

I advise ownCloud users to follow directly this issue for ownCloud integration of wallabag: https://github.com/wallabag/wallabag/issues/408

**Comment by rnewman** [comment]
> Proposal: Support 1.1 until a simpler 1.5 version protocol and OwnCloud plugin is done/contributed.

We've been waiting two years. How much longer would you have us wait? If we didn't care about self-hosters, we'd have killed 1.1 support as soon as 1.5 entered the release channel.

> Bottom line is that if Firefox wants to seriously continue to promote Freedom for your users' data, it should make self-hosting sync implementations much easier…

I disagree with your absolutism. All of this is about trade-offs: we try to make *all* users' data secure and private, rather than just those privileged enough to be able to self-host. We're not hypocritical enough to offer defaults that we aren't willing to use ourselves.

Any time Mozilla contributors spend working on Sync is best spent working on the client implementation, not on an ownCloud plugin.

> What I feel is being ignored is the amount of user effort required to follow Mozilla's FxA/Sync 1.5 agenda…

Not ignored; weighted. A tiny number of users, many of whom are technically competent enough to follow a blog post and docs. Yes, we'd like to improve self-hosting (some, like callahad, are quite passionate about it); one day perhaps it'll even be one-click. But to suggest that we prioritize those users' experience over the security and experience of 400M other people — as we would be if we encouraged disabling updates — is ludicrous.

There's only one reasonable course of action here: contribute improvements to Sync 1.5's self-hosting docs, and work on building a new version of ownCloud's sync plugin. Anything else is fighting the inevitable. If you care about this, work to make it happen in any way you're able. This is free software: you will not get what you want by loudly asserting that your use case is the most important.

**Comment by allo-** [comment]
When we're talking about alternatives: Here is my project to share urls from android to an self-hosted webinterface (and vice versa with a bookmarklet for your browser):

https://github.com/allo-/django-sharedurllist
https://github.com/allo-/SharedUrlList

Disclaimer: I did not build the android app since android 4.4 (my binary still works), the django app is untested on newer django. pull-requests welcome, i guess a firefox extension would be easy to do with the new SDK (never got to implementing it, though).

**Comment by allo-** [comment]
@rnewman inevitable is nothing. Sync 1.1 works fine at the moment, i never saw the point in 1.5, IMHO it lowers security a bit at login and complicates the protocol (iirc it has a dumbed down interface as well).
On the other hand, sync 1.5 isn't that bad, but the protocol is too complicated to get sync servers that easy, as you see. For sync 1.1 there were soon serveral alternatives (i am still using the django one), for 1.5 people seem even to struggle getting the official one running.

But i guess a owncloud bug is no place do discuss firefox design decisions.

**Comment by nickel-chrome** [comment]
@allo- is right running the FxA account server is no simple feat. I've managed to do it for an end-to-end encryption project I have been working on, https://github.com/eXfio/CucumberSync, but it required a lot of work and as the FxA account server is written in nodejs I found I needed a virtual server with root access to host it.

@rnewman In addition to FxA would Mozilla consider supporting OAuth2 or Persona authentication for Firefox Sync?

I believe OAuth2 is on the roadmap for both the FxA team and the ownCloud community, but would require quite a bit of work.

Firefox Accounts
https://developer.mozilla.org/en-US/docs/Mozilla/Tech/Firefox_Accounts/Introduction#OAuth_2.0_API

ownCloud
https://github.com/owncloud/core/issues/10400

Although Persona is also being end-of-lifed by Mozilla, based on my understanding of Sync 1.5 this would be a relatively minor change to the protocol, namely:

1. Request BrowserID certificate from Persona IdP rather than FxA, e.g. http://owncloud.example.com/cert_key
   https://github.com/mozilla-services/android-sync/blob/74579df0bb4037cb94abaac4c24fdbe35a7477d6/src/main/java/org/mozilla/gecko/fxa/login/Cohabiting.java#L25-L49

2. Derive client state from sync key rather than kB, which is FxA specfic.
   https://github.com/mozilla-services/android-sync/blob/74579df0bb4037cb94abaac4c24fdbe35a7477d6/src/main/java/org/mozilla/gecko/background/fxa/FxAccountUtils.java#L172-L189

3. In token server, nothing to do?

4. In ownCloud implement a 'good enough' Persona IdP.

**Comment by rnewman** [comment]
> @rnewman In addition to FxA would Mozilla consider supporting OAuth2 or Persona authentication for Firefox Sync?

Persona: no. When one EOLs a technology, one typically doesn't build new things on top of it. 

OAuth2, again, almost certainly not. (Indeed, FxA provides OpenID Connect compatible OAuth flows, so the relationship is reversed.)

The point of this whole thread is that maintaining different pluggable auth mechanisms (and in these cases, different pluggable key sources!) is complexity that introduces bugs and maintenance costs.

> … a relatively minor change to the protocol…

There is no such thing as a minor change to a deployed protocol with three clients on three platforms. It is way more likely that someone will ship a virtual container pre-configured to run the FxA + Sync stack than it is that we'll change the protocol to avoid the need to run FxA.

**Comment by jknockaert** [comment]
@rnewman The "whatever doc" I quoted from happened to be published by a mozilla engineer on a mozilla blog. In it's opening statement it says:
>The ability to self-host is central to Mozilla’s mission, as it ensures that the convenience of Sync doesn’t come with hidden costs like strict lock-in.

What I wanted to say is that based on the plan as it was published (I properly referenced to that "whatever doc") it was a reasonable strategy for self-hosters to stick to the ownCloud implementation. Sync 1.1 is in fact no more than a hack as there was never a consistent synchronisation concept implemented; but it was better than nothing and it worked. I think many users of the ownCloud implementation (the solution that is "central to Mozilla's mission") would very much like to move along to Sync 1.5, but they're simply dumped.

**Comment by nickel-chrome** [comment]
@jknockaert In Mozilla's defence I think they've done a good job with the standalone syncserver for sync 1.5 and if you look in to the FxA authentication protocol you will find that Mozilla never knows the password of an FxA account and even with malicious intent could only discover an equivalent of a password hash. So if your password is strong you are as secure as you can be for username/password based authentication.

It's a shame that the FxA account server is not so easy to host. If hosting your own authentication server is important to you, and with other options ruled out, I believe @rnewman is right that the way forward would be a virtual container that could be spun up on the major IaaS providers, i.e. EC2, Google Compute, OpenStack. As it turns out I may be able to contribute to this as my hosting service dreamhost are pushing customers off VPS to an OpenStack solution.

**Comment by craigarno** [comment]
@nickel-chrome well said, "I believe @rnewman is right that the way forward would be a virtual container that could be spun up on the major IaaS providers...", but still missing at least one of the major points.

The point is to be able to run the sync protocol without an outside internet connection.  I have no problem with Mozilla servers or anyone elses server unless my single point of failure, my internet connection, is not functioning, for -any- reason, intentional or not... 

Maybe I'm running from a tiny 1-man/woman SOHO or a several hundred employee company.  I don't want sync, or worse, workstation browsers to break or become partially disabled just because I don't have access to someone else's server for authentication or sync.  Particularly in a 1-person SOHO with internal resources, a requirement to run a separate VM/machine just to have auth services for sync or Firefox/Thunderbird isn't acceptable.

There are other use cases where one might not have an internet connection, like a secure facility working on a project which forbids outside access while working on "the project", but has their own internal web based resources.  Cell phones and other communication devices checked with security prior to entering to ensure no external access.

I could go on, but until a PHP-FxA/Sync1.5 solution is available for OwnCloud (or FxA/Sync1.5 in a replacement for the services I use, currently provided by OwnCloud), I still think my best solution is to freeze Firefox updates so I can keep using Sync 1.1/Mozilla_Sync 1.4 with or without an internet connection in my current setup.

If Mozilla as an entity made a decision that Mozilla must drop Sync 1.1 from Firefox, there is little else which can be done, follow or take a different path.  If Mozilla is right, then the few of us who'll disappear won't matter much compared to the millions who don't care.  I'll do the usual things, like look for a browser which does support Sync 1.1 (Weave).  Names which immediately come to mind are Waterfox, Palemoon, etc.

Bottom line; the new credential and sync system has to be relatively easy to implement and work seamlessly without an external server before I can adopt the solution. (one -tiny- voice among the millions).

**Comment by boltronics** [comment]
And also doesn't address the issue whereby some hosting providers are limited in the back-end scripting languages they will support, and also do not support such containers.

There are way too many blockers to making Sync 1.5 a practical alternative to 1.1 for self-hosting. I'm in the same boat as @craigarno - if Sync 1.1 is dropped, I'll have to find a different browser. That is, unless Mozilla comes up with a usable PHP-FxA/Sync1.5 solution first.

**Comment by Larx** [comment]
Without being a programmer, I think the best solution would be to drop any "browser sync" compatibility completely, and develop the bookmarks app to something like Shaarli or SemanticShuttle or Wallabag. And then write a browser plug in that integrates these shared bookmarks seamlessly into Firefox, Chrome, etc.
However, easier said than done, I'm afraid.

**Comment by d98ama** [comment]
@Larx This assumes that you only care about bookmarks, I don't. I mostly use the sync for passwords/form fields.

Personally I would be fine with a pre-configured Docker image. Then you can run it on your own box, or in the cloud. It wouldn't be integrated with OwnCloud, but there isn't really much integration with todays mozilla_sync plugin and the rest of OwnCloud anyway...

**Comment by allo-** [comment]
I guess the biggest point for using owncloud is NOT the raspberri pi, the IAAS or the cheap vserver, but actually the free (or very cheap) php webspace somewhere. And there is NO solution planned for that.
And even if it would be an alternative, it's quite strange from mozilla to tell owncloud "hey, tell your users to use another product".

**Comment by allo-** [comment]
Isn't the syncserver more or less still a key-value-store? Mozilla could just add a (hidden?) option "use simple authentication", which requires username, password and encryption password and logins via http-auth to the sync server and encrypts stuff with the encryption password. all features not available without FxA can be left out, weave 1.1 features should be possible with such a simple solution. Would even be possible to do auth by ip, i.e. in a LAN.

**Comment by craigarno** [comment]
@allo- "*...strange from Mozilla to tell OwnCloud "hey, tell your users to use another product".*"

That was me saying I'd have to use another product for the interim, not Mozilla.

Actually Mozilla has been pretty open about their decision to leave Sync 1.1 as it was causing a lot of problems both technical and support on their own systems.  FxA/Sync 1.5 is their best stab at coming up with a more workable solution for themselves and the majority of their user-base.  Mozilla also talked about a Sync 2.0 after Sync 1.5.  Maybe 2.0 will work for all of us where FxA/Sync 1.5 doesn't right now.  There is hope for the future, but the future isn't here yet.  I view this discussion as the current FxA/Sync 1.5 isn't workable for some of us, so we'll have to find other solutions until we can *potentially* rejoin later.  Yeah, it's bumpy ride while Mozilla tries to grow and stay relevant to its users with limited resources.

I use all 6 sync data types; Add-ons, Bookmarks, Passwords, Preferences, History, and Tabs.  Add-ons and Tabs are the only two I don't use heavily, but would miss if they disappeared.

**Comment by allo-** [comment]
@craigarno As the post just before ... i guess the sync isn't a problem, as everything needs to be there in the browser and the server side can be very simple. The complicated FxA flows, which provide features which are for stuff like login to bugzilla, appstore, etc. as well and not only to sync, they are the complicated thing.

**Comment by craigarno** [comment]
I keep copies of Firefox 29 (x86) and Waterfox 28 (x64) to setup sync 1.1 on my new computers to join the rest of my Sync collection.  The only thing which may change for me is to keep a copy of the last version to support already setup Sync 1.1.  For the moment, "If it ain't broke, don't fix it." will work for me.  I'll probably keep using Firefox (maybe FirefoxESR) and Waterfox if I can figure out how to disable updates, and/or make time to search for other solutions.

**Comment by franzhartwig** [comment]
@boltronics - what browser do you suggest? I am syncing FF on Ubuntu, Win7 and Android, Bookmarks and Passwords with Owncloud and need an alternative solution, if Mozilla drops Sync 1.1.

**Comment by allo-** [comment]
If you need one device at a time:
- Always close firefox
- Put the whole profile into owncloud (maybe excluding the cache, but on recent firefoxes it is stored in ``~/.cache`` anyway).

Be careful, syncing a profile between two running firefoxes may break quite a lot of things.
OTOH ancient mozilla had a "sync profile to FTP-Server" function, maybe there is some firefox addon, which does it in a similiar way.

**Comment by franzhartwig** [comment]
@allo- of course, syncing the whole profile folder may be an option. But what about FF on Android? No way to access the profile folder ...

**Comment by allo-** [comment]
yep. There you will have bad luck with alternative browsers as well.

**Comment by craigarno** [comment]
@franzhartwig , since you ask...

Under Linux I run **Firefox x64**, whatever comes with the distribution.  Windows (7,8,10) x64 I use **Waterfox (x64)**. Android, so far **Firefox**, but I haven't been able to get any version of **Android Firefox** to enter the Sync 1.1 (weave) circle provided by OwnCloud.  Android (Setup->accounts->Firefox Sync-depreciated) before version 29 sync says enter "..." (literally periods), which obviously isn't going to work.  After v29, FF android sync is determined to make me use the new sync with Firefox servers, and the older (depreciated) Sync 1.1 in Android "Setup->Accounts" doesn't want to connect either.  I still think Android Sync 1.1 is possible, I just haven't figured out how yet.

This said, I setup *Sync 1.5* for someone non-technical (one of the other millions of users FF is hoping to attract).  I setup Waterfox x64 in their Windows 10 laptop, and setup Android Firefox sync 1.5 on their android phone.  It all went pretty seamlessly, just as Mozilla intends.  This user asked for help because MSIE is incredibly slow on their machine (minutes to start), and Chrome broke page printing with their new cloud based printer setup, and is also quite slow.  Waterfox by comparison is blinding fast on their machine.  This user has zero understanding of what "Sync" is, but was happy that everything on their phone matched what is on their desktop.  When I say zero understanding, I'm referring to comments like, "*well I don't want my bank account copied into my phone*", with similar comments being made about their email.  i.e. zero comprehension about what a browser application does on their computer or phone.  This is fine, this is representative of most people using the internet, and what Mozilla has to address for a user-base.

I'm an Engineer, I just use these tools to make my life easier while I design medical electronics and software.  i.e. I'm not an I.T. person, just another user with a particular set of skills.

**Comment by Larx** [comment]
I'm still regularly using ownCloud Mozilla sync on Android, without any problem?!?! You just have to add the "legacy" sync adapter or something like that, don't know the exact wording.
In fact, the ability to sync between desktop and mobile Firefox is for me the most important feature of Mozilla sync. Many other aspects can be worked around...

**Comment by rnewman** [comment]
@Larx: the old account type is being removed in Firefox 46 for Android.

@craigarno Mozilla's J-PAKE pairing servers were shut down, so the legacy account pairing flow won't work, hence the endless "…".

**Comment by Larx** [comment]
I know that it will be removed with Version 46, that is what this thread is about, however it still works for me now and has not been broken since v29. Thats what I wanted to say.

**Comment by craigarno** [comment]
@rnewman what about custom Sync 1.1 accounts which have account/password/Custom-serverURL/Recovery-Key?  Will this information still work for adding Sync 1.1 ( FFv28/v29) installations without Mozilla servers?  Or were Mozilla servers always *silently* required for Sync 1.1 setup?

**Comment by allo-** [comment]
They will work. The only thing, which was hard to replace were the servers for coupling without recovery key.

**Comment by coder9** [comment]
Maybe it would be possible to take the code from the latest version of the old Weave/Sync plug-in that we can find, and bring that up to date. The plug-in could also disable the built-in Sync if about:config can't do it.

As for Android, could inbuilt Sync 1.5 be disable and Sync 1.1 support be made a separate installable account plug-in?

**Comment by rnewman** [comment]
> As for Android, could inbuilt Sync 1.5 be disable and Sync 1.1 support be made a separate installable account plug-in?

No.

> Maybe it would be possible to take the code from the latest version of the old Weave/Sync plug-in that we can find, and bring that up to date.

Theoretically possible (for desktop only, of course), but it's a fool's errand: not only would it be terrifically time consuming, but it would require lots of expertise in Sync internals. And who knows how long it would continue to work? Sync is necessarily coupled to core Firefox, which is why we discontinued the Sync add-on in the first place.

It would be more work to build a Sync add-on than it would be to build a Sync 1.5 ownCloud plugin and a configuration add-on.

This line of inquiry is verging on crazy talk: "I really hate the interior in new Volkswagens, so I went down to the wrecking yard and got an old Volvo 850, a mid-90s Golf, and a large spool of thread…".

Recognize when you're paddling upstream.

> what about custom Sync 1.1 accounts which have account/password/Custom-serverURL/Recovery-Key? Will this information still work for adding Sync 1.1 ( FFv28/v29) installations without Mozilla servers? Or were Mozilla servers always silently required for Sync 1.1 setup?

The J-PAKE servers were always required for the pairing flow, even self-hosted. If you have all four pieces of info, you can use the "I'm not near another device" link. This is true prior to Firefox 46 on Android. Desktop already makes it difficult for you to sign in, and the UI is untested and could well be broken, but in theory still works.

**Comment by allo-** [comment]
What can you say about the needed extension apis (jetpack and the new web-extensions)? Will there be access to bookmarks (looks good) history and passwords, in a sense that a sync could be implemented (i.e. preserving the order of the bookmarks in the toolbar)?

I guess if the apis are there and easy to use, some "simple sync" extension would really be an option.

**Comment by rnewman** [comment]
Not really in my wheelhouse, I'm afraid. I don't know of any plans to expose those APIs on Android, and I wouldn't want to try to build a robust sync solution on most APIs of that sort — you need change tracking, lifecycle events, batch operations, transactions.

**Comment by MichaWen** [comment]
I've just updated Firefox for Android from 44 to 45 through Play Store.
After the update finished, my "Firefox Sync (outdated)" account has disappeared and Firefox showed me a form to create a firefox-account.

To put it plainly: This happened after updating from 44 to 45.

So be careful if you planning to freeze your Firefox Version in order to further use of sync 1.1 .

**Comment by rnewman** [comment]
I suspect you created a Firefox Account on desktop. Android respects the migration sentinel. The removal of the old account type only happened last week; there's no way it's in 45. See https://bugzilla.mozilla.org/show_bug.cgi?id=1220906 .

**Comment by MichaWen** [comment]
It definitely happened on my Android Phone.

After the update I recognized a Firefox-Logo in the notification tray and touched it without reading the additional text. 

I thought it was on of these notifications for receiving a link, shared through "send to another device" from my tablet.

Firefox has been opened and showed me a form to create a Sync account. Then I checked the Android  Settings / Accounts and "Firefox Sync (outdated)" was gone.

Meanwhile I was able to recover it using the "I'm not near another device" link and everything is ok for now.

**Comment by rnewman** [comment]
Yes, that's the old Sync -> FxA migration flow. That you were able to recover an old Sync account on the device proves that. Note that your old Sync account will indeed disappear in two releases, about 12 weeks.

**Comment by elraro** [comment]
@Larx how u make work mozilla_sync owncloud with android? i tried ALL, a lot of versions of firefox android and nothing...

**Comment by rnewman** [comment]
@elraro: you'll need a hosting setup that doesn't use SNI.

**Comment by Larx** [comment]
@elraro: I didn't do anything special, I just added a Firefox Sync (legacy) adapter on Android. My OC is self-hosted, and I retain some old Firefox version on my desktop devices in order to initially set up a profile with legacy sync on them.

**Comment by Larx** [comment]
With the update to FF45 on Android (other devices still on 44), the legacy sync started behaving weired (doubling bookmarks all the time etc.). On one Android device, the legacy sync adapter completely disappeared already, on another it was still present.
As there is no solution for self-hosting (one that a normal human being can set up) is available, I switched to Firefox's own servers :-( .

**Comment by franzhartwig** [comment]
@Larx: Giving Mozilla all your data is not nessesary. You can install your own syncserver, only authenticating with the Mozilla server:

Install dependencies:
`sudo apt-get install python-dev git-core python-virtualenv`

Compile and install syncserver:
```
cd /opt
sudo git clone https://github.com/mozilla-services/syncserver
cd syncserver
sudo make build
```

Generate random secret for syncserver:
`head -c 20 /dev/urandom | sha1sum`

Edit syncserver config:
```
sudo nano syncserver.ini
public_url = http://localhost:5000/ -> replace to real (DynDNS) address
sqluri = sqlite:////opt/syncserver/file.db
```
or, if you use mysql:
```
sqluri = pymysql://root:yourmysqlpassword@localhost:3306/syncserver
secret = yourrandomnumber
allow_new_users = true
``` (this should be changed to false later)

Create database:
```
mysql -u root -p
CREATE DATABASE syncserver;
GRANT ALL PRIVILEGES ON syncserver.* TO sync IDENTIFIED BY „password“;
exit;
```

Set ownerchip and link to pserve:
```
sudo chown -R pi:pi /opt/syncserver
sudo ln -s /opt/syncserver/local/bin/gunicorn /usr/bin/gunicorn
```

Test syncserver:
`gunicorn --paste /opt/syncserver/syncserver.ini`

Stop syncserver:
`Ctrl + C`

Create init-Script:

`sudo nano /etc/init.d/syncserver`
Paste the following:
```
#!/bin/bash

### BEGIN INIT INFO
# Provides: syncserver
# Required-Start: $all
# Required-Stop:
# Default-Start: 2 3 4 5
# Default-Stop: 0 1 6
# Short-Description: starts the Firefox Sync Server
# Description: Starts the Firefox Sync Server
### END INIT INFO

PROJECT=/opt/syncserver
VIRTUALENV=/opt/syncserver/local
PID_DIR=/opt/syncserver
PID_FILE=/opt/syncserver/syncserver.pid
LOG_FILE=/opt/syncserver/logs/syncserver.log
USER=pi
GROUP=pi
PROD_FILE=/opt/syncserver/syncserver.ini

source $VIRTUALENV/bin/activate

cd $PROJECT

case „$1“ in
start)
gunicorn --paste $PROD_FILE --daemon --log-file=$LOG_FILE \
--pid $PID_FILE --user=$USER --group=$GROUP
;;
stop)
kill cat $PID_FILE
;;
restart)
kill cat $PID_FILE
gunicorn --paste $PROD_FILE --daemon --log-file=$LOG_FILE \
--pid $PID_FILE --user=$USER --group=$GROUP

;;
*)
echo $“Usage: $0 {start|stop|restart}“
exit 1
esac

exit $RET_VAL
```

Install init files and start syncserver:
```
sudo update-rc.d syncserver defaults
sudo service syncserver start
```

Configuring logrotate for rotating the syncserver logs may be a good idea.
The log is written to /opt/syncserver/logs/syncserver.log

In Firefox, go to about:config and change the syncserver URI:
services.sync.tokenServerURI

**Comment by elraro** [comment]
there is a Firefox alternative with sync 1.1, Pale Moon

**Comment by Larx** [comment]
Thanks for the instructions, I'll look into that (can I delete my data from Mozilla's servers once I've done the switch)?
I've looked into Palemoon, too, but it lacks proper Android support.

**Comment by franzhartwig** [comment]
I did not sync my data to Mozilla - don't know, if it can be deleted. Maybe you should backup your Firefox profile, delete all synced data, sync to Mozilla. Disable your internet connection, restore your old profile, change services.sync.tokenServerURI and enable your internet connection.

**Comment by jknockaert** [comment]
@franzhartwig Thank you for posting instructions for setting up a syncserver. But what is actually the point in hosting the data yourself as long as an external party is controlling the access to that data?

**Comment by ckruetze** [comment]
@jknockaert I'm following this discussion too as I would love this to work with owncloud in the future.
As for your question while not perfect it would help us a lot if we host the data.
Our server is on a private network and clients connect to it through VPN. While I wouldn't want Mozilla to control access to my data, they can't access it them self without the VPN.

**Comment by jknockaert** [comment]
@ckruetze I have been thinking along similar lines where the syncserver would only be available on the local network (possibly with vpn access to that network). But what will you do when Mozilla pulls the plug on the authentication service? They've been pulling the plug on mostly everything they've started up over the last decade, including Firefox OS, the old Sync, XUL, Thunderbird development, etc. And I'm afraid that Firefox for Android may well be the next thing to be dumped, leaving not much to be synced anyway.

**Comment by franzhartwig** [comment]
@jknockaert Of course, it would be better to run the account server on my own. But what are the choices? You can stop all syncing. Or you can sync your data to Mozilla.
Using external account servers is a common way. On many websites you see something like "Login with Facebook", "Login with Google", "Login with Yahoo". It's OpenID, not an exotic thing. So authenticating with my Mozilla account and syncing to my own syncserver is not so bad. If there is an easy way to setup the account server, I'll do it. But as long you need nodejs and install with npm, it's a pain. I tried it, but I failed.

**Comment by Larx** [comment]
If it would be easy, somebody would probably have thought of it, but cannot the sync server be patched to accept owncloud authorization instead of Mozilla authentication?

**Comment by franzhartwig** [comment]
@Larx Maybe the syncserver can be patched to accept the owncloud authentication. But I am not able to read the code, not to mention to patch it.

**Comment by allo-** [comment]
I am looking a bit in the WebExtensions API, it looks like it would be not too complicated to write a sync addon with it, which has the simplicity of sync 1.1. The problem: Important stuff like the history api are not yet supported, which means there is a gap between the first possible own sync and sync 1.1. Disregarding the development time as well.
The example addons seem really easy, where much can be achived with little code and chrome like api (so the addon could be chrome/opera/vivaldi/firefox all in one).
I am not sure about the sync 1.1 data model, i think creating an own simpler protocol would be good. key-value store on server side, some objects storing which keys are bookmarks, etc. with timestamps, as list of deleted objects for syncing deletions and objects stored as client-side encrypted data (needs some encryption library which does not look insecure).

**Comment by marcelklehr** [comment]
@allo- I think your approach could work. I've started work on an Addon using the WebExtensions API that taps into this hole: Enter [floccus](https://github.com/marcelklehr/floccus)!

I invite anyone to join me as this seems something the browser vendors just won't start themselves.

The addon in its current state can only pull bookmarks from the Bookmarks app, but I intend to either push the people behind that app to pull the pull request that adds CRUD methods to the REST API, or adopt WebDAV as a syncing protocol.

As @allo- pointed out, Firefox does not yet support the History API, but it shouldn't be too long until that happens.

I'd also like to cc @marklindhout who has done some promising work on utilizing WebDAV for syncing Bookmarks.

**Comment by allo-** [comment]
Looks like a nice start.

I think must have is for many people some kind of client-side encryption, that's the important point of sync 1.1.
And if you're not implementing the sync 1.1 protocol, a nice (uncomplicated) API with good documentation would be good, so other implementations (i.e. not requireing a whole owncloud) will be possible.

I had here a few lines of code iterating bookmarks and windows/tabs, but are waiting on the history api. Personally i would prefer a very dumb server as key-value store:
- Full Data in key-value pairs (encrypted)
- Collections (just information which keys represent which objects)
- Computers: coupled computers, associated with collections
- Log: timestamped actions, especially "delete", for incremental sync and removing stuff which once was in a collection. add/delete/change, type (bookmark, history), item (key)
- Session: Data which is not stored long term, i.e. tabs open on other computers. A computer refreshes it regularily and it is possibly deleted if it's too old. Other options would be "send tab to pc" or some kind of shared clipboard. No log here, always full loading.

**Comment by craigarno** [comment]
I would also like to see independence from external servers for all phases, like authentication, initial setup, and operation. Or from past experience; I'd like independence from Mozilla servers with compatibility where desired.  Where this is coming from is my older MozillaSync 1.1/OwnCloud setup still works well for me as an already established user.  Setting up a "new user" is quite difficult because there are dependencies on Mozilla servers which no longer support Sync 1.1 services.  And Mozilla recently broke, then restored Sync 1.5 login because of recent attack activity.  I prefer not to find myself in this Mozilla server dependency pit again.  What Sync 1.1 does, it does very well once it's setup. I'd like the Add-on Sync to work on a local network (without internet access), or securely across the internet with only my OwnCloud resources and local computer.  I use MozillaSync 1.1 with my OwnCloud setup when I travel, or work off site.  At least one of my clients blocks all external traffic except HTTP and HTTPS, so the solution needs to be able to use these interfaces, and I see this proposed one [Floccus] does.

If I had secure MozillaSync 1.1 "like" services in an add-on for Firefox/Waterfox (adding Chrome is a plus as long as Google servers aren't involved) which works with my OwnCloud setup with client side encryption... then Mozilla's discontinuing Sync 1.1 support in Firefox should be a non-issue for me.

From what little I can see, Floccus looks good guys! (yes, I went and looked, don't know what to do with it, but I know how to read)

**Comment by marklindhout** [comment]
Hi all, just responding to the mention I got via @marcelklehr: what I have right now is a grunt-based workflow that allows one to build Firefox and Chrome plugins easily, and a set of platform-agnostic central JavaScript files that can talk to wrappers for each browser's http request, bookmarking, and settings -API's.

However, that is all based on the add-on frameworks the browser vendors offered about half a year ago, and at that time the WebExtensions API was not ready for production yet. (I have no idea as to it's current state though, it might well be ready)

I've mostly done the groundwork for very basic bookmark management: Settings page and saving, making WebDav requests, creating bookmarks and folders all work. However, there's no sync logic yet in place, no encryption in settings storage, no decided format for actually saving the bookmark data.

So, it's rather rough, but if you need to get up and running fast, my workflow can definitely help you out with that.

**Comment by allo-** [comment]
Chrome seems to have quite a bit of the API, firefox has this page as documentation for their progress:
http://arewewebextensionsyet.com/

One big step will be addUrl for history, with the possiblity to change the creation time, the type (typed url, visited link, etc.) and so on, see https://bugzilla.mozilla.org/show_bug.cgi?id=1265836

The extension discussed there will be firefox only at the moment, as the chrome/opera api just has the `url` parameter. Without the possibilty to set the creation time, a useful sync will be hard. I guess if this should be a chrome extension as well, you need to open a bug in the chromium project for the same feature.

**Comment by rnewman** [comment]
At the risk of sounding like a party pooper: the WebExtensions bookmarks API isn't transactional, so even if you safely use WebDAV transactions, you cannot safely and correctly synchronize structural bookmark changes using that API.

You can get part of the way there by implementing your own two-phase reads and writes, but I doubt you'd ever get to a point where you had no lingering bugs and/or some amount of potential data loss.

Beware also of undefined behavior around restoring bookmark backups, Refresh Firefox, etc. The import events don't seem fine-grained enough to me. Neither do the change events seem to be documented to fire on all kinds of changes.

Amusingly, Firefox Sync (née Weave) itself started many years ago as an add-on using Places bookmark APIs to synchronize bookmarks via WebDAV. WebDAV was ditched very early on for its truly atrocious scalability, and over time we've recognized that building a synchronization system that isn't intimately tied to the data layer is an exercise in frustration.

I simply would not attempt to build an offline-capable synchronization system without (a) transactions,  (b) very detailed change/lifecycle hooks, and (c) thorough documentation of the API guarantees.

**Comment by allo-** [comment]
That what i would address with the concept, i lied out roughly:
- Do a full constructive sync every now and then, using the whole collection/key-value-store. Do not delete anything, but restore (later: Setting how careful this should work, rather losing one item or rather restoring one incorrectly)
- sync on action: use the "bookmark added", "bookmark removed", "page visisted" events to push it in real time (maybe bundles every 30 seconds or something like this).
- incremental sync: use the log to get actions from other devices, compare to an own log, which pushed items really were stored.
- on inconstencies fall back to full sync again.

With very conservative deletion policy (the action log should be a good indicator, a missing data-item not), the system will be mostly self healing, even when something got lost in between.

**Comment by allo-** [comment]
By the way, you're all php folks? I am still using https://github.com/allo-/django-sync-server and could sketch some of the database and API design in django very fast, trying to provide a useful api, which can be implemented with other frameworks as well. But of course such a backend would not really fit into owncloud.

**Comment by marcelklehr** [comment]
First up, sorry for the delay on my part.

On the status of the addon:
I'm in the process of getting floccus approved for AMO, so that it can be tested. @rnewman I'm already employing a layer of indirection for syncing the bookmarks that holds only the intersection of client and server. The syncing algorithm covers everything other than nested bookmarks and client-side update (i.e. you can't change the URL/title, atm) This is a first test though, and I believe it is best to transition to a different backend than the ownCloud bookmarks app.

Something that I think would be a perfect fit to the client-side encryption requirement is [CRDTs](https://github.com/dominictarr/crdt): The server doesn't need to know _anything_, it just needs to store the encrypted changes to the data and the client would have no transformation overhead. For adding a device using PAKE to negotiate a session key for sharing the master key would be nice (thanks to @callahad for the suggestion).

Regarding a server API: @rnewman Thank you for your insight! What was so atrocious about WebDAV? I'm guessing every item (i.e. bookmarks, etc.) was treated as a WebDAV 'file'/resource? When using CRDTs, only a log is necessary, so I think the [Atom Publishing Protocol](https://web.archive.org/web/20070301043011/http://www-128.ibm.com/developerworks/library/x-atompp1/) would be a nice thing to utilize.

**Comment by allo-** [comment]
I think with different Clients and Servers in Mind, Encryption should not be too complicated. More like base64 encoded AES or something like this. Client-only is useful, coupling can be done by shared secret (random UUID?) or when both devices are live and logged in to the same account there could be some Key exchange like in Diffie-Hellman scheme.
Such stuff should be not too complicated, as crypto schemes made up by yourself usually fail. So some library for key exchange and encryption using most basic standard algorithms would be useful.

**Comment by callahad** [comment]
PAKE might be worth looking into for key exchange. For an applied example, check out Brian Warner's talk on his project, [Magic Wormhole](https://www.youtube.com/watch?v=tgve3Nb1Xjk), at PyCon last month.
